finality-aleph/src/sync/substrate/verification/cache.rs

```diff
@@ -0,0 +1,340 @@
+use std::{
+    collections::HashMap,
+    fmt::{Display, Error as FmtError, Formatter},
+};
+
+use aleph_primitives::BlockNumber;
+
+use crate::{
+    session::{first_block_of_session, session_id_from_block_num, SessionId},
+    session_map::AuthorityProvider,
+    sync::substrate::verification::{verifier::SessionVerifier, FinalizationStatus},
+    SessionPeriod,
+};
+
+/// Ways in which a justification can fail verification.
+#[derive(Debug, PartialEq, Eq)]
+pub enum CacheError {
+    UnknownSession,
+    SessionTooOld(SessionId, SessionId),
+    SessionInFuture(SessionId, SessionId),
+}
+
+impl Display for CacheError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), FmtError> {
+        use CacheError::*;
+        match self {
+            SessionTooOld(session, lower_bound) => write!(
+                f,
+                "justification from too old session {:?}. Should be at least {:?}",
+                session, lower_bound
+            ),
+            SessionInFuture(session, upper_bound) => write!(
+                f,
+                "justification from session {:?} without known authorities. Should be at most {:?}",
+                session, upper_bound
+            ),
+            UnknownSession => write!(f, "justification from unknown session"),
+        }
+    }
+}
+
+/// Cache storing SessionVerifier structs for multiple sessions. Keeps up to `cache_size` verifiers of top sessions.
+/// If the session is too new or ancient it will fail to return a SessionVerifier.
+/// Highest session verifier this cache returns is for the session after the current finalization session.
+/// Lowest session verifier this cache returns is for `top_returned_session` - `cache_size`.
+pub struct VerifierCache<AP, FS>
+where
+    AP: AuthorityProvider<BlockNumber>,
+    FS: FinalizationStatus,
+{
+    sessions: HashMap<SessionId, SessionVerifier>,
+    session_period: SessionPeriod,
+    finalization_status: FS,
+    authority_provider: AP,
+    cache_size: SessionId,
+    /// Lowest curretnly available session.
```

Comment:
nit: documenting private fields feels weird :P

Reply:
Yeah, I know. I left this comment here as I had no better/more descriptive name for this field.